feat: note deduplication

benesjan opened the pull request:
Fixes F-93

Before this PR a note might have been returned multiple times from the `note_data_provider` because a note might have been present in multiple scopes and we didn't handle that. In this PR I fix this.

benesjan commented:
* **#18913** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/18913?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/18913?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `next`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

benesjan commented on yarn-project/pxe/src/storage/note_data_provider/note_dao.test.ts:
```diff
@@ -1,9 +0,0 @@
-import { NoteDao } from '@aztec/stdlib/note';
```
Forgot to move this test to stdlib when I was moving there `NoteDao`.

benesjan commented on yarn-project/stdlib/src/note/note.test.ts:
```diff
@@ -19,4 +19,12 @@ describe('note', () => {
     expect(jsonParseWithSchema(jsonStringify(note), Note.schema)).toEqual(note);
   });
+
+  it('equals returns false when number of items is different', () => {
```
There was a bug in the note's equals method where it didn't checked lengths and then it was comparing an item from note 1 with undefined ([this](http://ci.aztec-labs.com/a7b2fff97e4320b6) was the run).

mverzilli commented on yarn-project/pxe/src/storage/note_data_provider/note_data_provider.ts:
```diff
@@ -323,7 +324,15 @@ export class NoteDataProvider {
     }
 
-    return result;
+    // A note might be present in multiple scopes - we ensure it is only returned once
+    const deduplicated: NoteDao[] = [];
+    for (const note of result) {
+      if (!deduplicated.some(existing => existing.equals(note))) {
+        deduplicated.push(note);
+      }
+    }
```
Mega nit: if we make `result` a JS `Map<id, NoteDao>` maybe this code gets a bit more straightforward to read and write. Then again, wer'e probably going to torch this file soon so maybe not worth the hassle

benesjan replied:
Then we would need to convert the array back to array to return it from the function which made me not bother with it.

benesjan committed:
feat: note deduplication

Fixes F-93

Before this PR a note might have been returned multiple times from the `note_data_provider` because a note might have been present in multiple scopes and we didn't handle that. In this PR I fix this.
